packages/features/bookings/lib/handleNewBooking.ts

```diff
@@ -1984,11 +1984,46 @@ async function handler(
 
   try {
     if (hasHashedBookingLink && reqBody.hashedLink && !isDryRun) {
-      await prisma.hashedLink.delete({
+      const hashedLink = await prisma.hashedLink.findUnique({
         where: {
           link: reqBody.hashedLink as string,
         },
       });
+
+      if (hashedLink) {
+        let shouldDeleteLink = true;
+        const now = new Date();
+
+        if (hashedLink.expiresAt && hashedLink.expiresAt < now) {
+          throw new HttpError({ statusCode: 410, message: "Link has expired" });
+        }
+
+        if (hashedLink.maxUsageCount && hashedLink.maxUsageCount > 0) {
+          if (hashedLink.maxUsageCount > hashedLink.usageCount + 1) {
+            await prisma.hashedLink.update({
+              where: {
+                id: hashedLink.id,
+              },
+              data: {
+                usageCount: {
+                  increment: 1,
+                },
+              },
+            });
+            shouldDeleteLink = false;
+          } else {
+            shouldDeleteLink = true;
+          }
+        }
+
+        if (shouldDeleteLink) {
+          await prisma.hashedLink.delete({
```

Comment:
If hashed link is expired we never reach that code because with throw the error above https://github.com/calcom/cal.com/pull/20118/files#diff-a7c21727d9a0d4d1aaff9d24fca9293bc28468861bc8022dfa03116908b6fae7R1998